sql/core/src/main/scala/org/apache/spark/sql/execution/python/BatchEvalPythonExec.scala

```diff
@@ -17,18 +17,270 @@
 import org.apache.spark.api.python.{ChainedPythonFunctions, PythonRunner}
+import org.apache.spark.memory.{MemoryConsumer, TaskMemoryManager}
 import org.apache.spark.rdd.RDD
 import org.apache.spark.sql.catalyst.InternalRow
 import org.apache.spark.sql.catalyst.expressions._
 import org.apache.spark.sql.execution.SparkPlan
 import org.apache.spark.sql.types.{DataType, StructField, StructType}
+import org.apache.spark.unsafe.Platform
+import org.apache.spark.unsafe.memory.MemoryBlock
+import org.apache.spark.util.{CompletionIterator, Utils}
+
+
+/**
+ * A RowQueue is an FIFO queue for UnsafeRow.
+ */
+private[python] trait RowQueue {
+  /**
+   * Add a row to the end of it, returns true iff the row has added into it.
+   */
+  def add(row: UnsafeRow): Boolean
+
+  /**
+   * Retrieve and remove the first row, returns null if it's empty.
+   */
+  def remove(): UnsafeRow
+
+  /**
+   * Cleanup all the resources.
+   */
+  def close(): Unit
+}
+
+/**
+ * A RowQueue that is based on in-memory page. UnsafeRows are appended into it until it's full.
+ * Another thread could read from it at the same time (behind the writer).
+ */
+private[python] case class InMemoryRowQueue(page: MemoryBlock, fields: Int) extends RowQueue {
+  private val base: AnyRef = page.getBaseObject
+  private var last = page.getBaseOffset  // for writing
+  private var first = page.getBaseOffset  // for reading
+  private val resultRow = new UnsafeRow(fields)
+
+  def add(row: UnsafeRow): Boolean = {
+    if (last + 4 + row.getSizeInBytes > page.getBaseOffset + page.size) {
+      if (last + 4 <= page.getBaseOffset + page.size) {
+        Platform.putInt(base, last, -1)
+      }
+      return false
+    }
+    Platform.putInt(base, last, row.getSizeInBytes)
+    Platform.copyMemory(row.getBaseObject, row.getBaseOffset, base, last + 4, row.getSizeInBytes)
+    last += 4 + row.getSizeInBytes
+    true
+  }
+
+  def remove(): UnsafeRow = {
+    if (first + 4 > page.getBaseOffset + page.size || Platform.getInt(base, first) < 0) {
+      null
+    } else {
+      val size = Platform.getInt(base, first)
+      resultRow.pointTo(base, first + 4, size)
+      first += 4 + size
+      resultRow
+    }
+  }
+
+  def close(): Unit = {
+    // caller should override close() to free page
+  }
+}
+
+/**
+ * A RowQueue that is based on file in disk. It will stop to push once someone start to read
+ * from it.
+ */
+private[python] case class DiskRowQueue(path: String, fields: Int) extends RowQueue {
+  private var fout = new FileOutputStream(path)
+  private var out = new DataOutputStream(new BufferedOutputStream(fout))
+  private var length = 0L
+
+  private var fin: FileInputStream = _
+  private var in: DataInputStream = _
+  private val resultRow = new UnsafeRow(fields)
+
+  def add(row: UnsafeRow): Boolean = {
+    synchronized {
+      if (out == null) {
+        // Another thread is reading, stop writing this one
+        return false
+      }
+    }
+    out.writeInt(row.getSizeInBytes)
```

Comment:
Should these lines be inside of the `synchronized` body? I don't think that we're going to expect this lock to be super-contended since we only have one writer and one reader, so it might be simpler and more obviously-correct to just put the entire bodies of both `add`,  `remove`, and `delete` in `synchronized`.